### ISSUE TITLE
Upgrade to pac4j v2.0.0

### ISSUE THREAD

**Issue by leleuj** [ratpack/ratpack#1034]
Hi,

pac4j v1.9.1 is out and Ratpack should upgrade to use it.

The source code has been cleaned (-15% in size), dependencies (Java 8 also) have been upgraded, multi-profiles are supported and extension capabilities are much better -> https://github.com/pac4j/pac4j/wiki/Versions

Keeping the current Ratpack security algorithms and upgrading to pac4j v1.9.1 should be straightforward (most API signatures have remained the same). This is your option number 1: really easy!

---

Though, pac4j has reached a strong maturity with version 1.9: the security model is now available via specific components (and not only guidelines).
In addition to the concepts: clients, authorizers and matchers (whether to apply security or not), you now have three "filters":
1) The "security filter" (or whatever the mechanism used to intercept HTTP requests) protects an url by checking that the user is authenticated and that the authorizations are valid, according to the clients and authorizers configuration. If the user is not authenticated, it performs authentication for direct clients or starts the login process for indirect clients
2) The "callback controller" finishes the login process for an indirect client
3) The application logout controller" logs out the user from the application.

For example, for the "security filter", its logic is available via the [DefaultSecurityLogic](https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java) which makes implementations very easy:
- In J2E:

``` java
    @Override
    protected final void internalFilter(final HttpServletRequest request, final HttpServletResponse response,
                                        final FilterChain filterChain) throws IOException, ServletException {

        assertNotNull("securityLogic", securityLogic);

        final Config config = getConfig();
        assertNotNull("config", config);
        final J2EContext context = new J2EContext(request, response, config.getSessionStore());

        securityLogic.perform(context, config, (ctx, parameters) -> {

            filterChain.doFilter(request, response);
            return null;

        }, J2ENopHttpActionAdapter.INSTANCE, clients, authorizers, matchers, multiProfile);
    }
```
- In Play:

``` java
    public CompletionStage<Result> internalCall(final Context ctx, final String clients, final String authorizers, final boolean multiProfile) throws Throwable {

        assertNotNull("config", config);
        final PlayWebContext playWebContext = new PlayWebContext(ctx, sessionStore);
        final HttpActionAdapterWrapper actionAdapterWrapper = new HttpActionAdapterWrapper(config.getHttpActionAdapter());

        return securityLogic.perform(playWebContext, config, (webCtx, parameters) -> {
            // when called from Scala
            if (delegate == null) {
                return CompletableFuture.completedFuture(null);
            } else {
                return delegate.call(ctx);
            }
        }, actionAdapterWrapper, clients, authorizers, null, multiProfile, ctx);
    }
```

And the `ProfileManager` is the component to use to get the current authenticated user while the `SessionStore` is the abstraction for the web session.

As the pac4j calls are blocking (regarding asynchronous calls, it's another discussion), these "logic" calls should be wrapped in `Blocking.get()` or `Promise.async()` blocks.

The short guide: https://github.com/pac4j/pac4j/wiki/How-to-implement-pac4j-for-a-new-framework---tool

So this is your option number 2: a lot more work, but a lot more powerful: multiple clients / authorizers definitions, multi profiles authenticated at the same time...

---

What do you think?

Thanks.
Best regards,
Jérôme

**Comment by kyleboon** [comment]
I think ratpack should have different modules for the different pac4j versions. Maybe it is time to make ratpack-pac4j a separate project outside of the main repo?

**Comment by leleuj** [comment]
Currently, we have one dependency link: ratpack-pac4j pulls a certain version of pac4j. When upgrading the Ratpack version, sometimes the pac4j version is upgraded as well. This pac4j upgrade is for users very seamless because clients and authorizers configuration does not change (much) between versions (changes are easily detected via compilation failures).

The main issue is the Ratpack API to integrate with pac4j (like `RatpackPac4j.authenticator`): this must remain backward compatible as much as possible between versions.

I'm not sure to see the benefit of a separate module (which would depend on both Ratpack and pac4j).

Option 2 is more powerful and more consistent with other pac4j implementations, but it requires more work. It's a better choice because Ratpack would automatically benefit from the improvements developed in the pac4j core.
If we were to target this option, we certainly would have the following methods: `RatpackPac4j.security(config, clients, authorizers, matchers)`, `RatpackPac4j.callback(config, defaultUrl, logoutUrlPattern)`, `RatpackPac4j.logout(config, defaultUrl)`, `RatpackPac4j.profile(readFromSession)`... Existing methods would be deprecated and use these new ones.

So my recommendation would be to keep ratpack-pac4j as a submodule of Ratpack and go for option 2. I can of course help you in this upgrade providing you some basic pull request you could finish / improve.

And if Ratpack is to fully implement the pac4j security model, we should upgrade the demo (https://github.com/pac4j/ratpack-pac4j-demo) to make it the same as the other ones (like https://github.com/pac4j/j2e-pac4j-demo for example).

**Comment by leleuj** [comment]
I didn't see any update on this so I decided to submit a first PR for the simple upgrade to pac4j v1.9.1: https://github.com/ratpack/ratpack/pull/1048 (the demo has been upgraded as well: https://github.com/pac4j/ratpack-pac4j-demo/pull/8)

There should be no backward compatibility issues and the upgrade for Ratpack users should be straightforward.

I have an issue with the CAS support in the demo (authentication is restarted every time I access the /cas url): I'm certainly missing something obvious, but any insight would be appreciated.

When performing the upgrade and debugging into ratpack(-pac4j), I realized things were not easy due to two main factors: async nature of Ratpack and many concepts of pac4j. Plus I noticed that many things are duplicated (requested url erasure) or are missing (logs) in ratpack-pac4j while they are already / properly handled in pac4j.
So, to tackle option number 2 (all pac4j power in Ratpack), there is no other realistic solution than wrapping pac4j behaviours in blocking calls like we do in Play (or Vertx) for example:

```java
public CompletionStage<Result> callback() {

    assertNotNull("callbackLogic", callbackLogic);
    assertNotNull("config", config);
    final PlayWebContext playWebContext = new PlayWebContext(ctx(), playSessionStore);

    return CompletableFuture.supplyAsync(() -> {
        return callbackLogic.perform(playWebContext, config, config.getHttpActionAdapter(), this.defaultUrl, this.multiProfile, false);
    }, ec.current());
}
```

If you agree with this approach, I'm willing to do it.

**Comment by danveloper** [comment]
@leleuj -- we do currently wrap the [call to get the profile in a blocking operation](https://github.com/ratpack/ratpack/blob/master/ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticator.java#L66) ... Are there other parts that need to be blocking as well?

**Comment by leleuj** [comment]
In fact, pac4j has three "filters":

1) The ["**security filter**"](https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java) (or whatever the mechanism used to intercept HTTP requests) protects an url by checking that the user is authenticated and that the authorizations are valid, according to the clients and authorizers configuration. If the user is not authenticated, it performs authentication for direct clients (= web services) or starts the login process for indirect clients (= Facebook, CAS...)

2) The ["**callback controller**"](https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java) finishes the login process for an indirect client

3) The [**application logout controller**"](https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/engine/DefaultApplicationLogoutLogic.java) logs out the user from the application.

These filters relies on the logics provided by pac4j. Integrating pac4j is about creating these three "filters" and delegating to the logics:

- in J2E (https://github.com/pac4j/j2e-pac4j/blob/master/src/main/java/org/pac4j/j2e/filter/SecurityFilter.java#L62), it's a simple delegation

- in Play (https://github.com/pac4j/play-pac4j/blob/master/src/main/java/org/pac4j/play/java/SecureAction.java#L88), it's a delegation wrapped in a blocking call.

So the logics provided by the pac4j engine are all to be considered blocking and properly wrapped.

In addition, the retrieval of the user profile(s) is ensured by the `ProfileManager` (blocking again), but this could be replaced by a full async component.

**Comment by mnasyrov** [comment]
Hi! Could someone please to sum up a current state of upgrading to the latest pac4j?

**Comment by leleuj** [comment]
The latest 1.9.x version is the 1.9.6. As this new stream (1.9.x) has some breaking changes (compared to the 1.8.x stream), it cannot be dumped directly into a 1.4.x or 1.5.x version of Ratpack.

We talked about creating a separate project: `ratpack-pac4j`, hosted in the `ratpack` organization.

In May 2017, we will release the version 2.0 of pac4j which is a MAJOR one. We should focus our efforts on integrating it (instead of the 1.9.x stream).

**Comment by mkobit** [comment]
It sounds more like this issue is now "upgrade to pav4j v2.0.0", or is there plans to support both?

**Comment by leleuj** [comment]
Yes, as pac4j v2.0.0 has been released, `ratpack-pac4j` should use it.

@alkemist @danveloper: what are we doing? Either ratpack v1.5 accepts breaking changes by upgrading to pac4j v2.0 or a new `ratpack-pac4j` project is necessary.

I'm willing to create a `ratpack-pac4j` project in the `pac4j` organization (at first a copy of what exists in the `ratpack-pac4j` module of  `ratpack`) and we could deprecate the existing module in favor of this new project.

Sounds good?

**Comment by leleuj** [comment]
@alkemist @danveloper Are you in line that I create a `ratpack-pac4j` library on the `pac4j` organization side?

**Comment by KangoV** [comment]
So do I take it that ratpack will jump straight to 2.0.0? I was hoping we could have an interim upgrade to 1.9.x to fix some bugs like this one https://github.com/pac4j/pac4j/issues/500, which build the UserProfile with the wrong class which reults in a ClassNotFoundException.

**Comment by leleuj** [comment]
OK. It's time to do something for Ratpack and pac4j and I just created: https://github.com/pac4j/ratpack-pac4j

Is anyone willing to submit a first pull request to this new project with the current source code of the existing ratpack-pac4j module (https://github.com/ratpack/ratpack/tree/master/ratpack-pac4j)?

We'll upgrade to pac4j v1.9.8 and have a ratpack-pac4j (org.pac4j) version 1.5.0. I'll do the first release afterward.

We'll then create and release a ratpack-pac4j v2.0.0 based on pac4j v2.0.

**Comment by leleuj** [comment]
I've done the "copy/paste" and we have now two libraries working the same way: io.ratpack:ratpack-pac4j and org.pac4j:ratpack-pac4j.

The demo (https://github.com/pac4j/ratpack-pac4j-demo) uses the new library (org.pac4j:ratpack-pac4j).

I will release this library soon and then upgrade it to pac4j v2.0.

**Comment by leleuj** [comment]
I released ratpack-pac4j v2.0.0: https://github.com/pac4j/ratpack-pac4j

ratpack-pac4j v2.0.0 (groupId: org.pac4j) is the same library as the official ratpack-pac4j module (groupId: io.ratpack) but based on pac4j v2.1: it's almost fully backward compatible.
